app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationHelper.java

```diff
@@ -233,51 +234,135 @@ public static void make(final Context context, Notification body, AccountEntity
+            // Create a notification that summarises the other notifications in this group
 
-        notificationManager.notify(notificationId, builder.build());
-        if (currentNotifications.length() == 1) {
-            notificationManager.notify((int) account.getId(), builder.setGroupSummary(true).build());
-        } else {
-            notificationManager.notify((int) account.getId(), summaryBuilder.build());
+            // All notifications in this group have the same type, so get it from the first.
+            String notificationType = members.get(0).getNotification().extras.getString(EXTRA_NOTIFICATION_TYPE);
+
+            Intent summaryResultIntent = new Intent(context, MainActivity.class);
+            summaryResultIntent.putExtra(ACCOUNT_ID, (long) accountId);
+            summaryResultIntent.putExtra(TYPE, notificationType);
+            TaskStackBuilder summaryStackBuilder = TaskStackBuilder.create(context);
+            summaryStackBuilder.addParentStack(MainActivity.class);
+            summaryStackBuilder.addNextIntent(summaryResultIntent);
+
+            PendingIntent summaryResultPendingIntent = summaryStackBuilder.getPendingIntent((int) (notificationId + account.getId() * 10000),
+                pendingIntentFlags(false));
+
+            String title = context.getResources().getQuantityString(R.plurals.notification_title_summary, members.size(), members.size());
+            String text = joinNames(context, members);
+
+            NotificationCompat.Builder summaryBuilder = new NotificationCompat.Builder(context, channelId)
+                .setSmallIcon(R.drawable.ic_notify)
+                .setContentIntent(summaryResultPendingIntent)
+                .setColor(context.getColor(R.color.notification_color))
+                .setAutoCancel(true)
+                .setShortcutId(Long.toString(account.getId()))
+                .setDefaults(0) // So it doesn't ring twice, notify only in Target callback
+                .setContentTitle(title)
+                .setContentText(text)
+                .setSubText(account.getFullName())
+                .setVisibility(NotificationCompat.VISIBILITY_PRIVATE)
+                .setCategory(NotificationCompat.CATEGORY_SOCIAL)
+                .setOnlyAlertOnce(true)
+                .setGroup(channelId)
+                .setGroupSummary(true);
+
+            setSoundVibrationLight(account, summaryBuilder);
+
+            // TODO: Use the batch notification API available in NotificationManagerCompat
+            // 1.11 and up (https://developer.android.com/jetpack/androidx/releases/core#1.11.0-alpha01)
+            // when it is released.
+            notificationManager.notify(summaryTag, accountId, summaryBuilder.build());
+
+            // Android will rate limit / drop notifications if they're posted too
+            // quickly. There is no indication to the user that this happened
+            try { Thread.sleep(1000); } catch (InterruptedException ignored) { }
```

Comment:
Yeah, I think that documentation is wrong.

To prove this, do the following:

Modify `NotificationFetcher.fetchNewNotifications()`, and replace the block of code that calculates `minId` with this:

```kotlin
//        val minId = when (val marker = fetchMarker(authHeader, account)) {
//            null -> account.lastNotificationId.takeIf { it != "0" }
//            else -> if (marker.lastReadId > account.lastNotificationId) marker.lastReadId else account.lastNotificationId
//        }
        val minId = account.lastNotificationId
```

This will fetch new notifications based on your reading position in the Notifications tab.

Then comment out the `Thread.sleep` calls here and in `NotificationFetcher`.

Build and launch the app. Switch to the notifications tab, and scroll down a few pages, so that your reading position is before the newest notifications.

Switch back to the home tab. Then relaunch the app.

Tusky will re-fetch the Mastodon notifications that were above your reading position in the Notifications tab and display them as Android notifications.

In my emulator (running API 31) I see the following errors logged by the system when it does that (without the sleep):

```
NotificationService               E  Muting recently noisy 0|com.keylesspalace.tusky|2|147077155|10151
NotificationService               E  Muting recently noisy 0|com.keylesspalace.tusky|2|com.keylesspalace.tusky.notification.group_summary.CHANNEL_FAVOURITEmastodon.social:109241526707263377|10151
NotificationService               E  Package enqueue rate is 67.90989. Shedding 0|com.keylesspalace.tusky|2|148269585|10151. package=com.keylesspalace.tusky
```

If I'm reading that correctly it means that Android is both muting some notification channels because of the rate of new notifications (so users will not be informed about them) and dropping (i.e., "Shedding") some notifications entirely.

I suspect this is the source of some of the user complaints about missing notifications. If you get a lot in one go (say, a post has recently been boosted or favourited a lot) generating a lot of notifications then Android will mute and/or drop them.

That link says:

> Beginning with Android 8.1 (API level 27), apps can't make a notification sound more than once per second. If your app posts multiple notifications in one second, they all appear as expected, but only the first notification per second makes a sound.
> However, Android also applies a rate limit when updating a notification. If you post updates to a single notification too frequently, such as many in less than one second, the system might drop updates.

I don't think the first paragraph is relevant here.

The second paragraph should be relevant, and that's what I think is wrong. The code in this PR calls `.notify()` on each Android notification exactly once, so it's not "post(ing) updates to a single notification too frequently".

https://cs.android.com/android/platform/superproject/+/master:frameworks/base/services/core/java/com/android/server/notification/NotificationManagerService.java;l=6905-6937?q=%22Package%20enqueue%20rate%22 shows that this is checking `appEnqueueRate`.

And if you go poking around [frameworks/base/services/core/java/com/android/server/notification/NotificationUsageStats.java](https://cs.android.com/android/platform/superproject/+/refs/heads/master:frameworks/base/services/core/java/com/android/server/notification/NotificationUsageStats.java) it looks like this is the enqueue rate for all notifications, *not* a per-notification rate.

And if you look at https://cs.android.com/android/platform/superproject/+/refs/heads/master:frameworks/base/services/core/java/com/android/server/notification/NotificationManagerService.java;l=8129-8133;drc=7346c436e5a11ce08f6a80dcfeb8ef941ca30176 (where `isAlertRateLimited(pkg)` is called you) can see that the rate limits are definitely not calculated per notification, but across all the notifications posted by a single package (`pkg = record.getSbn().getPackageName()` -- `Sbn` is the status bar notification, `getPackageName()` is the name of the package that posted the notification).

https://cs.android.com/android/platform/superproject/+/refs/heads/master:frameworks/base/services/core/java/com/android/server/notification/AlertRateLimiter.java;l=23-35;drc=7346c436e5a11ce08f6a80dcfeb8ef941ca30176 is the rate limiter, which seems to set an allowed interval of 1 per 1000ms.

So I think Android applies undocumented rate limits to notifications, grouped by the package that posted them. Exceeding these limits will see notifications muted or dropped.